metricbeat/mb/testing/modules.go

```diff
@@ -212,15 +219,15 @@ func (r *CapturingReporterV2) GetErrors() []error {
 // ReportingFetchV2 runs the given reporting metricset and returns all of the
 // events and errors that occur during that period.
 func ReportingFetchV2(metricSet mb.ReportingMetricSetV2) ([]mb.Event, []error) {
```

Comment:
I'm afraid that I we shouldn't add context at all :sweat_smile: There are two main rules when implementing context:
- `context.Context` must be propagated from the initial call so `func ReportingFetchV2(metricSet mb.ReportingMetricSetV2) ([]mb.Event, []error)` must be `func ReportingFetchV2(ctx context.Context, metricSet mb.ReportingMetricSetV2) ([]mb.Event, []error)`. Propagation is key when using context.
- `context.Context` must not be stored anyhow. It's a continuation of the previous, if you follow the rule of propagation, you won't store it at all.

Finally, I don't see a real benefit of using context here. I mean that it looks like it's a "we can use Context here, as many other approaches" instead of a "it's a clear use case of Context". Also, I feel that we are replacing a piece with a explicit implementation which was already working. Explicit code is idiomatic in Go :) 

A not so old post about this: https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation

Reply:
> I'm afraid that I believe that we shouldn't add context at all :sweat_smile: There are two main rules when implementing context:
>     * `context.Context` must be propagated from the initial call so `func ReportingFetchV2(metricSet mb.ReportingMetricSetV2) ([]mb.Event, []error)` must be `func ReportingFetchV2(ctx context.Context, metricSet mb.ReportingMetricSetV2) ([]mb.Event, []error)`. Propagation is key when using context.
> 
>     * `context.Context` must not be stored anyhow. It's a continuation of the previous, if you follow the rule of propagation, you won't store it at all.
> 

In general I agree with that, but I'd like to avoid adding yet another interface (two of them if adding the one returning an error, plus its testing related code). We'd have to evaluate if it worths to add more interfaces for this.

Also this is a pattern also used in the standard library, for example [HTTP handlers](https://golang.org/pkg/net/http/#Handler) receive a [Request](https://golang.org/pkg/net/http/#Request) that has a `Context()`.

The context is only being stored where it is created, in the wrapper. We can continue with future refactorizations to replace `Start(done <-chan struct{})` with `Start(ctx context.Context)` so we have a context that can be propagated from there.

> 
> Finally, I don't see a real benefit of using context here. I mean that it looks like it's a "we can use Context here, as many other approaches" instead of a "it's a clear use case of Context". Also, I feel that we are replacing a piece with a explicit implementation which was already working. Explicit code is idiomatic in Go :)
> 

Well, most of our modules make requests to other services and many client libraries accept contexts.
The main benefit is that we can start using a context that makes sense in the current module lifecycle instead of using `Background()`, `TODO()` or avoiding APIs with contexts.

It is not using it "as many other approaches", it is using it as the approach that afaik is more widely used in the standard and third party libraries (even if we can discuss if it is the best approach).

> A not so old post about this: https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation

Nice read, but I think that start using something like `tomb` would require a huge effort in the case of Metricbeat, while getting a context from a _done channel_ we already have is quite straightforward. Also many libraries we use expect contexts, if we use other things we still need to convert them.

Reply:
When the first version of Beats was written, Context did not exist yet and we started with the `done` channel. Would we build it again I would assume Context would play quite a role in it. Long term I think more and more parts should be switched over to context but we need to go baby steps here (like in this PR). At least that is my thinking.